src/ngraph/runtime/cpu/pass/cpu_memory_optimization.cpp

```diff
@@ -0,0 +1,215 @@
+///                       \                 /
+///                               concat
+///
+/// After optimization: the result of add1 is stored to the memory buffer assigned to concat, same for add2 and add3.
+
+#include <cassert>
+
+#include "ngraph/runtime/cpu/pass/cpu_memory_optimization.hpp"
+
+#include "ngraph/descriptor/output.hpp"
+#include "ngraph/graph_util.hpp"
+#include "ngraph/op/concat.hpp"
+#include "ngraph/op/constant.hpp"
+#include "ngraph/runtime/cpu/cpu_op_annotations.hpp"
+#include "ngraph/runtime/cpu/mkldnn_utils.hpp"
+
+using namespace ngraph;
+
+bool runtime::cpu::pass::CPUMemoryOptimization::run_on_function(std::shared_ptr<Function> function)
+{
+    for (auto n : function->get_ordered_ops())
+    {
+        if (auto concat = std::dynamic_pointer_cast<op::Concat>(n))
+        {
+            auto shape = concat->get_input_shape(0);
+            auto axis = concat->get_concatenation_axis();
+            auto product = 1;
+            for (int i = 0; i < axis; i++)
+            {
+                product *= shape[i];
+            }
+            if (product != 1)
+            {
+                NGRAPH_DEBUG << "cpu_memory_optimization: The product of Concat's shape "
+                                "before concat axis is not 1, no in place concat";
+                continue;
+            }
+
+            bool in_place_concat = true;
+            AxisVector axis_list;
+            for (auto i = 0; i < shape.size(); i++)
+            {
+                axis_list.push_back(i);
+            }
+            auto index = 0;
+            for (descriptor::Input& input : concat->get_inputs())
+            {
+                // no tensors with zero-sized dimensions after zero_dim_tensor_elimination
+                assert(shape_size(input.get_shape()) != 0);
+
+                // check if input layout is padded
+                auto input_md = mkldnn_utils::get_input_mkldnn_md(n.get(), index);
+                index++;
+                if (mkldnn_utils::is_mkldnn_padded_layout(input_md, axis_list))
+                {
+                    NGRAPH_DEBUG
+                        << "cpu_memory_optimization: padded input layout, no in place concat";
+                    in_place_concat = false;
+                    break;
+                }
+
+                const auto& output = input.get_output();
+                auto arg = output.get_node();
+                if (std::dynamic_pointer_cast<op::Constant>(arg) ||
+                    std::dynamic_pointer_cast<op::Parameter>(arg))
+                {
+                    NGRAPH_DEBUG << "cpu_memory_optimization: " << arg->get_name()
+                                 << ": constant or parameter, no in place concat";
+                    in_place_concat = false;
+                    break;
+                }
+
+                if (arg->get_output_size() != 1)
+                {
+                    NGRAPH_DEBUG << "cpu_memory_optimization: " << arg->get_name()
```

Comment:
We seem to be running this pass before GOEE? So you shouldn't be seeing any multi-output ops feeding directly into `Concat`. You could change it to `NGRAPH_ASSERT`